xep-0388.xml

```diff
@@ -195,71 +300,187 @@
-      <p>In general, implementors are advised that a non-empty authorization string MAY be considered an error if the stream's from attribute (if present) does not match.</p>
-    </section2>
-    <section2 topic="Aborting">
-	    <p>Clients MAY abort unilaterally by sending &lt;abort/> as specified in <link url="#abort">Client Aborts</link>.</p>
+<section1 topic="SASL mechanism upgrade tasks" anchor="upgrade">
+  <p>Modern XMPP server deployments typically store only the hash of a user's password, to improve account security. At times, it may be desirable for servers to upgrade to newer or different hash algorithms, e.g. so they can offer different authentication mechanisms for improved security or interoperability. Due to the security properties of one-way cryptographic hash algorithms, it is not possible for the server to extract the original data and simply hash it in a new format. To perform such upgrades, the cooperation of the client is necessary - because it has, or can obtain from the user, the original password to derive a hash from.</p>
+  <p>As described in the <link url="#feature">SASL mechanism negotiation</link> section, the server can inform the client of one or more upgrades that need to be performed for the current account. The client SHOULD always request any upgrade task it recognises. To do this, it includes the 'upgrade' attribute in its &lt;authenticate/> element naming the upgrade task it wants to perform, as specified in <link url="#initiation">Initiation</link>.</p>
+  <p>Upon successfully authenticating the client (including any secondary authentication steps required for the account), but before the final &lt;success/> would be sent, the server sends a &lt;continue/> element, which MUST contain a single task, matching whatever was selected by the client. The client then initiates this upgrade task by providing a corresponding &lt;next/> element providing the task name.</p>
+  <p>Upon receiving the &lt;next/> element for the upgrade, the server provides the elements and data needed for the client to calculate the requested data. The concrete elements and exchanges needed for the upgrade are specific to individual tasks. This document currently defines tasks for upgrading credentials between different members of the SCRAM family of mechanisms. Other tasks may be documented in future versions of this document, or in other documents.</p>
+  <p>Upgrade task names SHOULD have a prefix of "UPGR-" followed by the SASL mechanism name to upgrade to, and if multiple mechanisms differ only in their support for channel-binding (e.g. SCRAM's -PLUS variants), implementations MUST use only the names of variants without channel-binding for the task names, because mechanism upgrades are independent of any channel-binding. Finally, upgrade tasks MUST NOT transmit plaintext passwords (or any reversible encoding of them) if the SASL mechanism to upgrade allows this to be avoided.</p>
```

Comment:
Every time you impose a restriction on the string syntax of an opaque identifier, a layer-violation fairy dies.

Less prosaically, either a client knows how to upgrade to a mechanism or else it doesn't, and either way the "UPGR-" doesn't really help.

Reply:
I wanted to make sure these tasks could eventually be used as SASL mechanism without introducing name clashes. I used `UPGR-` instead of `UPGRADE-` to keep the name below the 20 chars limit of SASL.
(`UPGR-` and `-PLUS` have the same size, and `-PLUS` won't be used for upgrades.

But I don't have a strong opinion about that.